esx_service/vmdk_ops.py

```diff
@@ -1670,22 +1683,44 @@ def set_vol_opts(name, tenant_name, options):
 
     return False
 
+def wait_ops_in_flight():
+    global WAIT_STOP_COUNT
+
+    # Set the STOP flag to true
+    stopEvent.set()
+
+    # Wait until the in-flight ops counter goes down to zero
+    while ((opsCounter.value > 0) and (WAIT_STOP_COUNT > 0)):
+        logging.info("Service is being stopped. Waiting for completion of %d operations in flight...", opsCounter.value)
+        time.sleep(WAIT_STOP_INTERVAL)
```

Comment:
I think this is the  major issue here... it should no wait in signal handler.

Reply:
Actually we should wait in signal handler.

As you also mentioned, once the signal handler takes over the main thread, the logic in the handleVmciRequests main loop will be hung. So if we move this wait into handleVmciRequests() - no matter it's in the loop or after the loop - it will never get a chance to run, because the logic in handleVmciRequests() will (almost) always gets stuck on the listening socket call - lib.vmci_get_one_op(). I have already verified this.

However, blocking on lib.vmci_get_one_op() has no impact to the system exit(0) call, which is triggered from the signal handler.

The logic is quite complicated here - I hope I have explained it clearly :)

Reply:
wait, who suggested to move waiting logic into handleVmciRequests()  ? 

I think the right approach is as follows: 
- signal handler either exit() right away (if count is 0) or sets a barrier and spawns a waiting thread
- waiting thread waits on an avent and when someone signals on the event, exits
- count.decr() signals on the event when barrier is set and count drops to 0

It is fairly standard drain-end-exit steps.

Reply:
Ah, too many comments got messed up:) Your suggestion above makes sense. I will try it out. Thanks!